### ISSUE TITLE
Valid JS numbers are not valid for this library

### ISSUE THREAD

**Issue by aklinker1** [MikeMcl/big.js#134]
For example:

```js
Number.parseFloat("1..")        // -> 1
Number.parseFloat("1..1")       // -> 1
Number.parseFloat("1........1") // -> 1
```

~~TBH I have no idea why this behavior exists, but it does, so I would assume this library to either handle these odd cases the same as JS~~.

> EDIT: This behavior is clearly defined by the [MDN docs](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/parseFloat#Description) (it's still weird though). If this library changed to support this type of parsing, it would most likely be a breaking change. See my second suggestion below for a better, non-breaking solution to this issue.

My second suggestion would be to expose way to check if a string is a valid number (since this library appears to do a better job than the standard JS library...).

For example, this will throw an error, when the regular JS way would return something, even if it's not right:

```js
new Big("1").add("1..1")                           // -> Invalid number
Number.parseFloat("1") + Number.parseFloat("1..1") // -> 2
```

**TLDR; Regular JS checks are not enough to clean inputs before sending them into this library.**

---

<details><summary>JavaScript man 🤣</summary>
<p>

![Javascript man](https://pics.me.me/drunk-member-this-is-my-favorite-javascript-language-redcoders-11-62100635.png)

</p>
</details>

**Comment by benfletcher** [comment]
>  so I would assume this library to either handle these odd cases

*Why* should someone assume that Big would coerce strings in the same way that a different numeric type does? (I have ideas, but you didn't give reasons beyond having had a wrong assumption.)

I think there are more people who would be surprised at Number's behavior than Big's failure to mimic that surprising behavior.

> OR EVEN BETTER expose way to check if a string is a valid number

It is exposed as your example demonstrates: it throws an error which can be caught in situations where the input to Big could be dirty.

And this is coming from someone who argued a few years ago that `Big('+1')` shouldn't throw an error (for which I still think an argument could made), so I certainly understand where you are coming from in this issue.

If coercion from strangely formatted number strings to Big is important, I'd argue it should live in a separate package whose sole purpose is cleaning the input. Otherwise it feels like we're mixing two distinct needs when most users won't need both things. Finally, the change would be breaking as you have to assume that at least some users of Big are relying on it throwing an error on dirty input.

**Comment by aklinker1** [comment]
@benfletcher I understand what you're saying. I've done a bit more research into why JS behaves like this, the behavior is clearly defined by [official MDN documentation for `parseFloat`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/parseFloat#Description). I definitely agree that supporting this kind of parsing behavior would be a very bad idea and breaking change. I will edit the original description to strike through that suggestion.

That said, I still consider this to be an issue.

> Why should someone assume that Big would coerce strings in the same way that a different numeric type does?

Because the documentation doesn't specify that it is different than the standard library. The [constructor](http://mikemcl.github.io/big.js/#big) does not mention this difference, and FAQs even state that you can use `parseFloat` to go from a `Big` &rarr; `number`. Without putting much more thought into it, it's a relatively easy (while wrong) assumption to make that the standard library should be able to sanitize inputs since this library does not provide an obvious alternative.

At least, I think that it would be worth noting in the docs that you shouldn't do something like this to check if your string is a valid input for this library:

```ts
function isNumberValid(input: string): boolean {
  return !isNaN(parseFloat(input));
}
```

> It is exposed as your example demonstrates

Yes, it does expose that something is wrong, but I wouldn't consider this a great work around for simple UI checks before doing any computations:

```ts
function isNumberValid(input: string): boolean {
  try {
    new Big(input);
    return true;
  } catch (err) {
    return false;
  }
}
```

If I'm going to use this on a front end to warn a user that an input is bad, actually creating the instance of the `Big` is useless since it won't be used. All the code needs to know is if the input is ready to be used with this library, then it can perform the needed computations when a form is submitted or something like that. Right now, this is the exact function we've written as a work around, but it feels like overkill.

> If coercion from strangely formatted number strings to Big is important, I'd argue it should live in a separate package whose sole purpose is cleaning the input.

As mentioned at the beginning of this message, I 100% agree that we shouldn't change the parsing. However, I don't think an additional helper method to perform these simple checks is beyond the scope of this library.

**Comment by benfletcher** [comment]
You could also mimic Big's [internal regex test](https://github.com/MikeMcl/big.js/blob/7c990e67aa3b49219d94a3f696b81769c72a5ae8/big.js#L68).

I like the idea of exposing that as a Big method after reading your explanation.

Thanks for the long explanation. It makes a lot more sense to me and was interesting to hear the reasoning.

**Comment by MikeMcl** [comment]
I am not sure why you focus on `parseFloat` rather than the the more natural analogue to the `Big` constructor: the `Number` constructor.

A validity check can be performed using `isFinite(Number())`, for example

```js
const isValid = n => isFinite(Number(n));
```

but unfortunately this would fail with inputs such as `"+2"` as the `Big` constructor doesn't accept the leading plus sign as mentioned by Ben above and at #59 and #70. 

Why not accept it? Because I don't want the check for it to have to be performed every time a Big number is created from a string. The `"+"` is redundant information and it is easy for the small minority of users who have values that include it to strip it out themselves, along with leading and trailing whitespace etc.

Should this library include a validity checking method? Possibly, see #61, but then the `Big` constructor already does this itself....

**Comment by aklinker1** [comment]
@benfletcher I didn't want to make it a super long read, so I tried keeping it short, but maybe it was a little too short. I like the idea of using the regex, but I don't like that there is the potential that that regex could change without me knowing about it (probably very unlikely at this point), as well as some of the other logic before the regex test is done.

@MikeMcl I agree `isFinite(Number(n))` is a better solution than `parseFloat`, I had no idea either of those functions existed to be honest. I still don't like that it isn't the same check done internally, so there still could be instances where an input is "valid" by that check but the constructor still throws an error like you pointed out.

I can understand why you don't want to add another function to the API at this point (I'm assuming you decided against adding the `test` method since that issue was from 2016), so I'll go ahead and close this issue and continue to try catch the constructor since that appears to be the safest way to validate inputs for now.
